wgengine/router/dns.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package router
+
+import (
+	"time"
+
+	"inet.af/netaddr"
+)
+
+// DNSConfig is the subset of Config that contains DNS parameters.
+type DNSConfig struct {
+	// Nameservers are the IP addresses of the nameservers to use.
+	Nameservers []netaddr.IP
+	// Domains are the domains to whose subdomains the DNS configuration applies
+	// if it is not global (e.g. under systemd-resolved).
```

Comment:
The former interpretation is accurate: resolved _only_ really supports per-domain resolvers because it does not have a user-settable priority. The interface with the default IP route will have DNS priority greater or equal to `tailscale0` in all cases, which makes claiming the global resolver status rather pointless: we will necessarily race against those other nameservers.

There are possible workarounds for this: for instance, as far as per-domain resolvers go, the most specific domain wins, so we could effectively become the global resolver by associating ourselves with all currently active TLDs, which should win over the root domain. Alternatively, we could mess with the other interfaces' settings and remove the DNS default route from them, but I'm reluctant to do that without dire need.

This makes resolved unpleasantly different from the other modes, to the extent that perhaps it should be considered its own separate thing. At the same time, it works perfectly for Magic DNS specifically, so I will just conditionally use it only in that case for now.